### PR TITLE
docs: add template ref var to glossary

### PR DESCRIPTION
**Pull request by kapunahelewong**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

Adds definition for "template reference variable" to the glossary.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
Per request by @jbogarthyde

### CODE REVIEW

File: aio/content/guide/glossary.md [hunk 1]
```diff
@@ -912,6 +912,12 @@ A TypeScript-like syntax that Angular evaluates within a [data binding](#data-bi
 
 Read about how to write template expressions in  [Template expressions](guide/template-syntax#template-expressions).
 
+## template reference variable
+
+A template reference variable, `#var`, is a reference to a DOM element within a template.
+
```
Comment (jbogarthyde) [inline aio/content/guide/glossary.md:918]: a reference anchor that you add to a DOM element within a template. Elements can share information by referring to one another through these variables.

(Maybe show how you make a reference using the variable?)

File: aio/content/guide/glossary.md [hunk 2]
```diff
@@ -912,6 +912,12 @@ A TypeScript-like syntax that Angular evaluates within a [data binding](#data-bi
...
+For more information about template reference variables, see the [Template reference variable](guide/template-syntax#ref-var) section of [Template Syntax](guide/template-syntax).
+
```
Comment (jbogarthyde) [inline aio/content/guide/glossary.md:920]: of the [Template Syntax](guide/template-syntax) guide.

**Review by jbogarthyde — COMMENTED** [review]
Maybe a little more complete?
Is this an Angular feature, or just HTML? If it's HTML, perhaps something about how it is used in Angular.

File: aio/content/guide/glossary.md [hunk 3]
```diff
@@ -912,6 +912,12 @@ A TypeScript-like syntax that Angular evaluates within a [data binding](#data-bi
 
 Read about how to write template expressions in  [Template expressions](guide/template-syntax#template-expressions).
 
+## template reference variable
```
Comment (jbogarthyde) [inline aio/content/guide/glossary.md:915]: add anchor: {@a template-reference-variable}

File: aio/content/guide/glossary.md [hunk 4]
```diff
@@ -912,6 +912,12 @@ A TypeScript-like syntax that Angular evaluates within a [data binding](#data-bi
...
+
+A template reference variable, `#var`, is a reference to a DOM element within a template.
+
+For more information about template reference variables, see the [Template reference variable](guide/template-syntax#ref-var) section of [Template Syntax](guide/template-syntax).
```
Comment (jbogarthyde) [inline aio/content/guide/glossary.md:919]: For more information, see...

**Review by jbogarthyde — APPROVED** [review]
LGTM

**Review by aikithoughts — APPROVED** [review]
Judy says good to go, so who am I to argue? :D

File: aio/content/guide/glossary.md [hunk 5]
```diff
@@ -919,6 +919,18 @@ A TypeScript-like syntax that Angular evaluates within a [data binding](#data-bi
 
 Read about how to write template expressions in  [Template expressions](guide/template-syntax#template-expressions).
 
+{@a template-reference-variable}
+
+## template reference variable
+
+A template reference variable, `#var`, is a reference to a DOM element within a template.
```
Comment (kara) [inline aio/content/guide/glossary.md:926]: These variables can also refer to component or directive instances, so we should probably reflect that in the definition somewhere. Maybe...:

```suggestion
A template reference variable, `#var`, is a reference to a DOM element, component instance, or directive instance within a template.
```

We should probably also explain the basic rules, i.e. that 
- if the var is declared on a component, it's the component instance
- if the var is declared on a normal tag, it's the element
- if the var has a specific export (e.g. `#var="ngModel"`), it's whichever directive or component matches that `exportAs` string

**Review by kara — DISMISSED** [review]
Notes on the glossary

File: aio/content/guide/glossary.md [hunk 6]
```diff
@@ -919,6 +919,24 @@ A TypeScript-like syntax that Angular evaluates within a [data binding](#data-bi
 
 Read about how to write template expressions in  [Template expressions](guide/template-syntax#template-expressions).
 
+{@a template-reference-variable}
+
+## template reference variable
+
+A template reference variable, `#var`, is a reference to a DOM element, component instance, or directive instance within a template.
+Elements can share information by referring to one another through these variables.
+In the following example `#phone` is the template reference variable.
```
Comment (jelbourn) [inline aio/content/guide/glossary.md:928]: The following example defines a template reference variable named `#phone`

(active voice)

File: aio/content/guide/glossary.md [hunk 7]
```diff
@@ -919,6 +919,24 @@ A TypeScript-like syntax that Angular evaluates within a [data binding](#data-bi
...
+
+<code-example path="template-reference-variables/src/app/app.component.html" region="ref-var" header="src/app/app.component.html"></code-example>
+
+There are several basic rules regarding template reference variables.
```
Comment (jelbourn) [inline aio/content/guide/glossary.md:932]: I would phrase this like
```markdown
The following semantics describe the behavior of template reference variables:
```

To me, the word "rules" implies they're restrictions on how the feature may be used, but we're actually just describing how the feature functions.

File: aio/content/guide/glossary.md [hunk 8]
```diff
@@ -919,6 +919,24 @@ A TypeScript-like syntax that Angular evaluates within a [data binding](#data-bi
...
+
+* If you declare the variable on a component, the variable refers to the component instance.
+* If you declare the variable on a standard HTML tag, the variable refers to the element.
+* If the variable has a specific export, for example, #var="ngModel", the variable refers the directive or component that matches that `exportAs` string.
```
Comment (jelbourn) [inline aio/content/guide/glossary.md:936]:
```suggestion
* If the variable specifies a name on the right-hand side (e.g. `#var="ngModel"`), the variable refers the directive or component on the element with a matching `exportAs` name.
```

**Review by IgorMinar — DISMISSED** [review]
+1 for @jelbourn's comments, except that I suggest that we move the second part of the description (semantics of the usage) into the template doc. This content is in my opinion not suitable for the glossary which should contain only a minimial description of a term with a reference to some other document providing more details (if necessary).

Reviewed-for: public-api

File: aio/content/guide/glossary.md [hunk 9]
```diff
@@ -919,6 +919,18 @@ A TypeScript-like syntax that Angular evaluates within a [data binding](#data-bi
 
 Read about how to write template expressions in  [Template expressions](guide/template-syntax#template-expressions).
 
+{@a template-reference-variable}
+
+## template reference variable
+
+A template reference variable, `#var`, is a reference to a DOM element, component instance, or directive instance within a template.
```
Comment (pkozlowski-opensource) [inline aio/content/guide/glossary.md:926]: We are missing the `<ng-template #ref>` in this list - in this case we are getting `TemplateRef` which is neither DOM element nor directive / component instance: https://stackblitz.com/edit/angular-azphjr?file=src/app/app.module.ts

File: aio/content/guide/template-syntax.md [hunk 10]
```diff
@@ -1777,6 +1776,12 @@ Here, a `<button>` further down the template refers to the `phone` variable.
 
 <code-example path="template-reference-variables/src/app/app.component.html" region="ref-phone" header="src/app/app.component.html"></code-example>
 
+The following semantics describe the behavior of template reference variables:
```
Comment (pkozlowski-opensource) [inline aio/content/guide/template-syntax.md:1779]: We are missing the `<ng-template #ref>` in this list - in this case we are getting `TemplateRef` which is neither DOM element nor directive / component instance: https://stackblitz.com/edit/angular-azphjr?file=src/app/app.module.ts

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
We need to cover the `<ng-template #ref>` case - see in-line comments.

File: aio/content/guide/glossary.md (hunk 9 above)
Comment (jelbourn) [reply aio/content/guide/glossary.md:926]: How about something like
```markdown
A variable defined in a template that references a directive instance,
components instance, DOM element, or embedded template.
```
or, if we don't want to be strictly comprehensive
```markdown
A variable defined in a template that references an instance
associated with an element, such as a directive instance, 
component instance, or DOM element.
```

I also noticed looking at this now the the start of the entry was inconsistent with other entries; it looks like most of the other entries don't start with "_term_ is..."

File: aio/content/guide/template-syntax.md (hunk 10 above)
Comment (jelbourn) [reply aio/content/guide/template-syntax.md:1779]: @kapunahelewong I suggest adding another bullet under the "standard HTML tag" point
```markdown
* If you declare the variable on an `<ng-template>` element, the variable refers to the template (represented by a `TemplateRef` instance).
```

File: aio/content/guide/glossary.md (hunk 9 above)
Comment (kapunahelewong) [reply aio/content/guide/glossary.md:926]: Added to this list. And thank you for the example! I'll add it to the example code in https://github.com/angular/angular/pull/31195

File: aio/content/guide/template-syntax.md (hunk 10 above)
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md:1779]: Thank you both! Added!

File: aio/content/guide/glossary.md (hunk 9 above)
Comment (jelbourn) [reply aio/content/guide/glossary.md:926]: Did you push your changes? I see the same text as earlier

Comment (pkozlowski-opensource) [reply aio/content/guide/glossary.md:926]: Yeh, I also don't see TemplateRef mentioned here. Is it intentional?

File: aio/content/guide/glossary.md [hunk 11]
```diff
@@ -919,6 +919,18 @@ A TypeScript-like syntax that Angular evaluates within a [data binding](#data-bi
...
+Elements can share information by referring to one another through these variables.
```
Comment (pkozlowski-opensource) [inline aio/content/guide/glossary.md:927]: Hmmm, this doesn't sound quite right to me. It is not like elements share information here.
I would rather say that directives can access to referenced things.

File: aio/content/guide/template-syntax.md [hunk 12]
```diff
@@ -1777,6 +1776,13 @@ Here, a `<button>` further down the template refers to the `phone` variable.
...
```
Comment (pkozlowski-opensource) [inline aio/content/guide/template-syntax.md:1779]: "semantics describe the behavior" - this reads wired to me, I'm not sure what it means... we just want to say that ref variables point to different things depending on where those are defined. Can we try to express it using simpler / more direct language?

Comment (jelbourn) [reply aio/content/guide/template-syntax.md:1779]: How about 
```markdown
Angular assigns each template reference variable a values based on where you declare the variable:
```

File: aio/content/guide/glossary.md (hunk 11 above)
Comment (jelbourn) [reply aio/content/guide/glossary.md:927]: How about
```markdown
After declaring a template reference variable on an element in a template,
you can access values from that variable elsewhere within the same template.
```

File: aio/content/guide/glossary.md (hunk 9 above)
Comment (kapunahelewong) [reply aio/content/guide/glossary.md:926]: Oh! I think I accidentally added it to another branch! Now I've put it in here. And I have added `TemplateRef` to the list. Is this ok?

> A variable defined in a template that references an instance associated with an element, such as a directive instance, component instance, template as in `TemplateRef`, or DOM element.

@pkozlowski-opensource I added a section on TemplateRef based on your example and notes, too: https://github.com/angular/angular/pull/31195/files#diff-fe1da8d4b293de490dd0cef57e6d036bR1880

Once that's in I can add a link, What do you two think?

**Review by jelbourn — APPROVED** [review]
LGTM

Reviewed-for: public-api

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM after latest edits, thnx!

Reviewed-for: public-api

**Comment by jelbourn** [comment]
@IgorMinar I think this one is just waiting on you now

**Review by aikithoughts — APPROVED** [review]
Reviewed-for: global-docs-approvers

**Comment by kapunahelewong** [comment]
I moved the edits to `template-syntax.md` into `template-reference-vars.md` since the PR to break apart that doc landed.

**Comment by jelbourn** [comment]
@atscott can you take a look for public-api?

**Review by atscott — APPROVED** [review]
reviewed-for: public-api

**Review by petebacondarwin — APPROVED** [review]
Reviewed-for: public-api

**Review by alxhub — APPROVED** [review]
Reviewed-for: public-api

**Pull request closed without merging**